File: LinkedList.java
public class LinkedList<T> {
    
    private Node<T> header;
    private Node<T> tail;
    private int count;
    
    public LinkedList() {
        header = null;
        tail = null;
        count = 0;
    }
    
    public void add(T elem) {
        Node<T> newNode = new Node<T>(elem);
        if (isEmpty()) {
            header = newNode;
            tail = newNode;
        } else {
            tail.nextNode = newNode;
            tail = newNode;
        }
        count++;
    }
        
    
    public T remove(int index) {
        if (index < 0 || index >= count) {
            throw new IndexOutOfBoundsException("Index: " + index + ", Size: " + count);
        }
        
        T removedValue;
        
        if (index == 0) {
            removedValue = header.getValue();
            header = header.nextNode;
            if (header == null) {
                tail = null;
            }
        } else {
            Node<T> current = header;
            for (int i = 0; i < index - 1; i++) {
                current = current.nextNode;
            }
            
            removedValue = (T) current.nextNode.getValue();
            
            if (current.nextNode == tail) {
                tail = current;
            }
            current.nextNode = current.nextNode.nextNode;
        }
        
        count--;
        return removedValue;
    }
    
    public T get(int index) {
        if (index < 0 || index >= count) {
            throw new IndexOutOfBoundsException("Index: " + index + ", Size: " + count);
        }
        
        Node<T> current = header;
        for (int i = 0; i < index; i++) {
            current = current.nextNode;
        }
        return current.getValue();
    }
    
    public boolean isEmpty() {
        return count == 0;
    }
    
    public int size() {
        return count;
    }
    
    public void print() {
        Node<T> current = header;
        System.out.print("LinkedList: ");
        while (current != null) {
            System.out.print(current.getValue() + " ");
            current = current.nextNode;
        }
        System.out.println();
    }
    
    public int contains(T elem) {
        Node<T> current = header;
        int index = 0;
        while (current != null) {
            if (current.getValue().equals(elem)) {
                return index;
            }
            current = current.nextNode;
            index++;
        }
        return -1;
    }
    
    public boolean removeFirst(T elem) {
        if (isEmpty()) {
            return false;
        }
        
        if (header.getValue().equals(elem)) {
            remove(0);
            return true;
        }
        
        Node<T> current = header;
        while (current.nextNode != null) {
            if (current.nextNode.getValue().equals(elem)) {
                if (current.nextNode == tail) {
                    tail = current;
                }
                current.nextNode = current.nextNode.nextNode;
                count--;
                return true;
            }
            current = current.nextNode;
        }
        
        return false;
    }
    
    public void insert(int pos, T elem) {
        if (pos < 0 || pos > count) {
            throw new IndexOutOfBoundsException("Index: " + pos + ", Size: " + count);
        }
        
        if (pos == count) {
            add(elem);
            return;
        }
        
        Node<T> newNode = new Node<T>(elem);
        
        if (pos == 0) {
            newNode.nextNode = header;
            header = newNode;
            if (tail == null) {
                tail = newNode;
            }
        } else {
            Node<T> current = header;
            for (int i = 0; i < pos - 1; i++) {
                current = current.nextNode;
            }
            newNode.nextNode = current.nextNode;
            current.nextNode = newNode;
        }
        count++;
    }
}

File: Linked Lists vs Array/linkedlist.py
from node import Node

class LinkedList:
    def __init__(self):
        self.header = None
        self.tail = None
        self.count = 0

    def add(self, elem):
        newNode = Node(elem)
        if self.isEmpty():
            self.header = newNode
            self.tail = newNode
        else:
            self.tail.nextNode = newNode
            self.tail = newNode
        self.count += 1

    def remove(self, index):
        if index < 0 or index >= self.count:
            raise IndexError("Index out of bounds")

        # Removing first element
        if index == 0:
            removedValue = self.header.getValue()
            self.header = self.header.nextNode
            if self.header is None:
                self.tail = None
        else:
            current = self.header
            for _ in range(index - 1):
                current = current.nextNode

            removedValue = current.nextNode.getValue()

            if current.nextNode == self.tail:
                self.tail = current

            current.nextNode = current.nextNode.nextNode

        self.count -= 1
        return removedValue

    def get(self, index):
        if index < 0 or index >= self.count:
            raise IndexError("Index out of bounds")

        current = self.header
        for _ in range(index):
            current = current.nextNode

        return current.getValue()

    def isEmpty(self):
        return self.count == 0

    def size(self):
        return self.count

    def output(self):
        current = self.header
        print("LinkedList:")
        while current is not None:
            print(str(current.getValue()) + " ", end="")
            current = current.nextNode
        print()

    def contains(self, elem):
        current = self.header
        index = 0
        while current is not None:
            if current.getValue() == elem:
                return index
            current = current.nextNode
            index += 1
        return -1

    def removeFirst(self, elem):
        if self.isEmpty():
            return False

        # First element
        if self.header.getValue() == elem:
            self.remove(0)
            return True

        current = self.header
        while current.nextNode is not None:
            if current.nextNode.getValue() == elem:
                if current.nextNode == self.tail:
                    self.tail = current
                current.nextNode = current.nextNode.nextNode
                self.count -= 1
                return True
            current = current.nextNode

        return False

    def insert(self, pos, elem):
        if pos < 0 or pos > self.count:
            raise IndexError("Index out of bounds")

        if pos == self.count:
            self.add(elem)
            return

        newNode = Node(elem)

        if pos == 0:
            newNode.nextNode = self.header
            self.header = newNode
            if self.tail is None:
                self.tail = newNode
        else:
            current = self.header
            for _ in range(pos - 1):
                current = current.nextNode
            newNode.nextNode = current.nextNode
            current.nextNode = newNode

        self.count += 1
